presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CombineApproxPercentileFunctions.java

```diff
@@ -0,0 +1,289 @@
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(CombineApproxPercentileFunctions::hasMultipleApproxPercentile);
+
+    private static boolean hasMultipleApproxPercentile(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream().filter(agg -> agg.getCall().getDisplayName().equals("approx_percentile")).count() > 1;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isCombineApproxPercentileEnabled(session);
+    }
+
+    private static int getPercentileArgIdx(List<RowExpression> arguments)
+    {
+        // approx_percentile(x, percentage) -> arguments.size() == 2
+        // approx_percentile(x, percentage, accuracy) -> arguments.size() == 3 && arguments.get(1) is Double
+        // approx_percentile(x, w, percentage) -> arguments.size() == 3 && arguments.get(1) is BigInt
+        // approx_percentile(x, w, percentage, accuracy) -> arguments.size() == 4
+        if (arguments.size() == 2 || (arguments.size() == 3 && (arguments.get(1).getType() instanceof DoubleType))) {
+            return 1;
+        }
+        else if (arguments.size() == 4 || (arguments.size() == 3 && (arguments.get(1).getType() instanceof BigintType))) {
+            return 2;
+        }
+        return -1;
+    }
+
+    private static Optional<List<RowExpression>> getNewpPercentileArgs(List<RowExpression> arguments, RowExpression newPercentileArg)
+    {
+        int percentileIdx = getPercentileArgIdx(arguments);
+        if (percentileIdx == -1) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<RowExpression> newAggCallArguments = ImmutableList.builder();
+        for (int i = 0; i < arguments.size(); i++) {
+            if (i == percentileIdx) {
+                newAggCallArguments.add(newPercentileArg);
+            }
+            else {
+                newAggCallArguments.add(arguments.get(i));
+            }
+        }
+        return Optional.of(newAggCallArguments.build());
+    }
+
+    private static Optional<AggregationNode.Aggregation> aggregationWithNewFunction(AggregationNode.Aggregation aggregation, Optional<CallExpression> call)
+    {
+        if (call.isPresent()) {
+            return Optional.of(new AggregationNode.Aggregation(
+                    call.get(),
+                    aggregation.getFilter(),
+                    aggregation.getOrderBy(),
+                    aggregation.isDistinct(),
+                    aggregation.getMask()));
+        }
+
+        // Only change the percentile argument to compare for match
+        RowExpression dummy = constant(0.0, DOUBLE);
```

Comment:
Hmm this is confusing. Can we do it in a better way?

Reply:
Refactor the code a bit, so that I directly remove the percentile argument to compare, rather than adding a dummy argument here.